### PR TITLE
Doc: Add directives, roles for declaring/referencing options

### PR DESCRIPTION
**Pull request by dbaston**
## What does this PR do?

This PR:
* creates new directives for defining config options, open options, dataset/layer creation options, including available choices, default values, and version introduced
* creates roles for referencing each of the above, or inserting a copy of the entire definition
* adds a config option to suppress "since" messages below a threshold version
* copies the global config definitions from the wiki (https://trac.osgeo.org/gdal/wiki/ConfigOptions) into the Sphinx docs

This ensures that config options are formatted consistently and allows them to be cross-referenced, e.g.

```
.. config:: GDAL_HTTP_MERGE_CONSECUTIVE_RANGES
   :since: 2.3
   :choices: YES, NO
   :default: NO
 
   Only applies when :config:`GDAL_HTTP_MULTIRANGE` is YES. Defines
   if ranges of a single ReadMultiRange() request that are consecutive should be
   merged into a single request. 
```

Errors are produced if a referenced config option does not exist or if multiple definitions of the same option exist in different files.

It also reproduces the "list of configuration options and where they apply" and has this list link back to the definition of each option.

## Tasklist

-  [x] Reformat config options on driver pages
-  [x] Handle PDF/text output
-  [x] Retire `decl_configoption`
 - [x] Add documentation
 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

### CODE REVIEW

File: doc/source/user/configoptions.rst [hunk 1]
```diff
@@ -142,6 +142,496 @@ Example:
...
+
+.. config:: CPL_TMPDIR
+   :choices: <dirname>
+
+   By default, temporary files are written into current working directory.
+   Sometimes this is not optimal and it would be better to write temporary files
+   on bigger or faster drives (SSD).
+
+
+.. config:: CPL_VSIL_CURL_ALLOWED_EXTENSIONS
```
Comment (rouault) [inline doc/source/user/configoptions.rst:199]: Doesn't fit in a "Logging" section (same for CPL_VSIL_CURL_NON_CACHED). Perhaps moved to the "HTTP section" which could be renamed "Networking"  ?

**Review by rouault — COMMENTED** [review]
Did you check the PDF output ?

**Comment by dbaston** [comment]
> Did you check the PDF output ?

It doesn't build (there's an entry for it on the Tasklist).

Just trying to get high-level feedback for now to see if the general direction seems OK. cc @hobu

**Comment by dbaston** [comment]
PDF output is fixed and the list of config options is reproduced.

I will make a full pass over the changeset to make sure I didn't accidentally mess up content with the reformatting. I will also flag spots where I made significant content additions/changes beyond reformatting.

**Comment by sgillies** [comment]
Awesome!

File: doc/source/user/virtual_file_systems.rst [hunk 2]
```diff
@@ -57,6 +57,25 @@ To use the .zip as a directory, you can use :file:`/vsizip/path/to/the/file.zip`
 
 Note: in the particular case where the .zip file contains a single file located at its root, just mentioning :file:`/vsizip/path/to/the/file.zip` will work.
 
+The following configuration options are specific to the /zip/ handler:
```
Comment (dbaston) [inline doc/source/user/virtual_file_systems.rst:60]: @rouault the various bulleted lists of options in this file are the only significant "new" content.

**Comment by dbaston** [comment]
I'm ready for this to go whenever you are.

File: doc/source/drivers/vector/eeda.rst [hunk 3]
```diff
@@ -34,58 +34,55 @@ Open options
...
-   value directly, it is the filename where the key should be read.
--  :decl_configoption:`EEDA_CLIENT_EMAIL` =string: email to be specified together with
-   EEDA_PRIVATE_KEY/EEDA_PRIVATE_KEY_FILE to use OAuth2 Service Account
-   authentication.
--  :decl_configoption:`EEDA_PAGE_SIZE`: Features are retrieved from the server by chunks 
-   of 1000 by default (and this is the maximum value accepted by the server). This number 
-   can be altered with this configuration option.
+-  :copy-config:`EEDA_BEARER`
+
+-  :copy-config:`EF EEDA_BEARER_FILE`
```
Comment (rouault) [inline doc/source/drivers/vector/eeda.rst:69]:
```suggestion
-  :copy-config:`EEDA_BEARER_FILE`
```

File: doc/source/drivers/vector/elasticsearch.rst [hunk 4]
```diff
@@ -385,120 +446,212 @@ Layer creation options
...
+      :choices: <bytes>
+      :default: 5000000
+
+      Identifies the maximum size in bytes of the
+      buffer to store documents to be inserted at a time. Lower record
+      counts help with memory consumption within Elasticsearch but take
+      longer to insert. Starting with GDAL 2.1, the :lco:`BULK_SIZE` layer
+      creation option should be used instead.
+
+-  .. config:: ES_OVERWRITE
```
Comment (rouault) [inline doc/source/drivers/vector/elasticsearch.rst:650]:
```suggestion
-  .. config:: ES_OVERWRITE
      :choices: YES, NO
      :default: NO
```
```

File: doc/source/drivers/vector/geoconcept.rst [hunk 5]
```diff
@@ -81,76 +81,85 @@ The OGR GeoConcept driver does not support deleting features.
...
+      ``TYPE`` corresponds to one of the ``Name`` found in the GCT file for a
+      type section. The ``SUBTYPE`` corresponds to one of the ``Name`` found
+      in the GCT file for a sub-type section within the previous type section.
 
-   At the present moment, coordinates are written with 2 decimals for
-   Cartesian spatial reference systems (including height) or with 9
-   decimals for geographical spatial reference systems.
+      At the present moment, coordinates are written with 2 decimals for
+      Cartesian spatial reference systems (including height) or with 9
+      decimals for geographical spatial reference systems.
```
Comment (rouault) [inline doc/source/drivers/vector/geoconcept.rst:162]: this remark doesn't seem related to the above lco
```suggestion
At the present moment, coordinates are written with 2 decimals for
Cartesian spatial reference systems (including height) or with 9
decimals for geographical spatial reference systems.
```

File: doc/source/drivers/vector/idb.rst [hunk 6]
```diff
@@ -41,11 +41,25 @@ Driver capabilities
...
+
+-  .. config:: DB_LOCALE
+
+      Locale of Informix database
+
+-  .. config:: CLIENT_LOCALE
+
+      Client locale
+
+-  .. config:: IDB_OGR_FID
```
Comment (rouault) [inline doc/source/drivers/vector/idb.rst:60]: IDB_OGR_FID is actually the only GDAL configuration option. The 4 above do not appear in GDAL sources, and must thus be real environment variables of the informix lib

**Review by rouault — COMMENTED** [review]
just a few correction suggestions. I tried to skim through everything but obviously not checking everything in detail. I can imagine your effort and pain to do all that editing ! Thanks again !!!

File: doc/source/drivers/vector/idb.rst (hunk 6 above)
Comment (dbaston) [reply doc/source/drivers/vector/idb.rst:58]: This prompted me to grep out all of the options from the docs and check them against the source; there are about about a dozen that I don't immediately see. I'll look into the following:

GDAL_VRT_TRUSTED_MODULES
GDAL_PDF_DPI
GDAL_PDF_NEATLINE
GDAL_USER_PWD
GDAL_PDF_BANDS
GDAL_PDF_LAYERS
GDAL_PDF_LAYERS_OFF
CARTO_PAGE_SIZE
LIBKML_USE_SCHEMADATA
LIBKML_BOTTOMTFOV_FIELD
LIBKML_NEAR_FIELD
LIBKML_SHAPE_FIELD
VSIS3_CHUNK_SIZE
VSIOSS_CHUNK_SIZE

Comment (rouault) [reply doc/source/drivers/vector/idb.rst:58]: > GDAL_VRT_TRUSTED_MODULES

should be GDAL_VRT_PYTHON_TRUSTED_MODULES. cf ../frmts/vrt/vrtderivedrasterband.cpp

> GDAL_PDF_DPI
> GDAL_PDF_NEATLINE
> GDAL_PDF_BANDS
> GDAL_PDF_LAYERS
> GDAL_PDF_LAYERS_OFF

dynamically requested from open options. cf "grep -r alt_config_option ../frmts/pdf"

> GDAL_USER_PWD

there must be some mismatch. It should be PDF_USER_PWD if I trust "alt_config_option='PDF_USER_PWD'/>"

> VSIS3_CHUNK_SIZE
> VSIOSS_CHUNK_SIZE
dynamically generated. cf ../port/cpl_vsil_s3.cpp:            (std::string("VSI") + poFS->GetDebugKey() + "_CHUNK_SIZE").c_str(),

> CARTO_PAGE_SIZE

cf ogr_carto.h

> LIBKML_USE_SCHEMADATA

no idea about that one

> LIBKML_BOTTOMTFOV_FIELD

typo, should be LIBKML_BOTTOMFOV_FIELD

> LIBKML_NEAR_FIELD

typo, should be LIBKML_NEARFOV_FIELD

> LIBKML_SHAPE_FIELD

perhaps LIBKML_PHOTOOVERLAY_SHAPE_FIELD

Comment (dbaston) [reply doc/source/drivers/vector/idb.rst:58]: LIBKML_USE_SCHEMADATA

no idea about that one

Looks like it should be `LIBKML_USE_SIMPLEFIELD` (c8e9b2ed519dd9d9ebcbf7cae0db1d56217c9eec)

**Comment by kannes** [comment]
Just noticed this due to https://trac.osgeo.org/gdal/wiki/ConfigOptions being moved.

Wow, great work! Thank you!
